In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import fetch_openml
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

# Loading in the MNIST Dataset

In [2]:
# I referenced the answer at https://stackoverflow.com/questions/47324921/cant-load-mnist-original-dataset-using-sklearn
# to figure out how to load the full size MNIST dataset using sklearn.
# Accessed 12/7/21
mnist = fetch_openml('mnist_784')

In [3]:
mnist_array = np.array(mnist.data) # Converting the data to a NumPy array

In [4]:
mnist_array = mnist_array.reshape((70000, 28 * 28)) # Vectorizing the images

In [5]:
mnist_labels = np.array(mnist.target, dtype='int32') # Turning the labels into an array of integers

In [6]:
x_train, x_test, y_train, y_test = train_test_split(mnist_array, mnist_labels) # Splitting training and testing data

In [7]:
clf = SVC() # Initializing an SVM

In [8]:
clf.fit(x_train, y_train) # Fitting the SVM on the training data

SVC()

In [9]:
preds = clf.predict(x_test) # Making predictions for the testing data

In [10]:
# Printing out performance metrics
print(classification_report(y_test, preds))
print(confusion_matrix(y_test, preds))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      1746
           1       0.98      0.99      0.99      1953
           2       0.97      0.98      0.97      1701
           3       0.97      0.96      0.97      1786
           4       0.98      0.98      0.98      1681
           5       0.98      0.98      0.98      1640
           6       0.99      0.99      0.99      1646
           7       0.98      0.98      0.98      1866
           8       0.98      0.97      0.97      1726
           9       0.98      0.97      0.97      1755

    accuracy                           0.98     17500
   macro avg       0.98      0.98      0.98     17500
weighted avg       0.98      0.98      0.98     17500

[[1731    2    1    1    1    1    5    0    4    0]
 [   0 1938    7    0    0    0    1    2    3    2]
 [   4    7 1662    6    1    1    2    9    6    3]
 [   0    1   20 1721    2   14    1    9   14    4]
 [   1    5    1    0 1650   

The SVM attains a 98% accuracy, which is very high. In this case, it seems unnecessary to do a hyperparameter search, as the default values give a model with about the highest accuracy we could reasonably expect, performing comparably to the MLP and almost as well as the CNN, which is impressive as it is a much simpler model.